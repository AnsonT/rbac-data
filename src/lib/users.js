import uuid from 'uuid/v4'
import _ from 'lodash-uuid'
import emailValidator from 'email-validator'
import { InvalidParameterError } from './errors'
import { ROOT_TENANT } from './constants'

export async function createUser (tx, { tenantId, userName, email, emailVerifiedAt, needNewPassword } = {}) {
  if (_.isUuid(userName)) {
    throw new InvalidParameterError({ message: 'Invalid parameter: userName' })
  }
  if (!emailValidator.validate(email)) {
    throw new InvalidParameterError({ message: 'Invalid parameter: email' })
  }
  userName = userName.toLowerCase()
  email = email.toLowerCase()
  const userId = uuid()
  const resp = await tx
    .insert({
      userId,
      userName,
      email,
      emailVerifiedAt,
      needNewPassword
    })
    .into('users')
  if (resp.rowCount === 1) {
    return { userId, userName }
  }
}

export async function listUsers (tx, { tenantId = ROOT_TENANT, offset = 0, limit = 20 }) {}

/**
 * Return a user by userId
 * @param {object} tx - Knex transaction object
 * @param {string} userId - the userId of the user
 */
export async function getUserById (tx, userId) {
  return tx
    .select()
    .from('users')
    .where({ userId })
    .first()
}

/**
 * Return a user based on userName or email (userName or email must be specified)
 * @param {object} tx - Knex transaction object
 * @param {object} where - the query parameters for the user
 * @param {string} where.tenantId - tenant realm of user, defaults to ROOT_TENANT
 * @param {string} where.userName - user name
 * @param {string} where.email - email
 */
export async function getUserByNameOrEmail (tx, { tenantId = ROOT_TENANT, userName, email }) {
  if (!userName && !email) {
    throw new InvalidParameterError({ message: 'userName or email must be specified' })
  }
  userName = userName && userName.toLowerCase()
  email = email && email.toLowerCase()
  tx = tx
    .select()
    .from('users')
    .where({ tenantId })
  if (userName) { tx = tx.where({ userName }) }
  if (email) { tx = tx.where({ email }) }
  return tx.first()
}

export async function removeUser (tx, userId) {}

export async function updateUser (tx, userId, { userName, email, emailVerifiedAt, needNewPassword }) {}
