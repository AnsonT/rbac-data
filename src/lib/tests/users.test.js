import _ from 'lodash'
import knex from './db'
import { createUser, getUserById, getUserByNameOrEmail, listUsers, removeUser, updateUser } from './../users'

const user1 = {
  userName: 'User1',
  email: 'user1@example.localhost',
  password: 'password'
}

describe('Users Tests', () => {
  beforeAll(async () => {
    await knex.migrate.rollback()
    await knex.migrate.latest()
    await knex.seed.run({ specific: 'unittests.js' })
  })
  afterAll(async done => {
    await knex.destroy()
  })

  it('listUsers', async (done) => {
    const omitTimestamps = (users) => _.map(users, user => _.omit(user, ['createdAt', 'modifiedAt']))
    let offset = 0
    const limit = 3
    while (true) {
      const users = omitTimestamps(await listUsers(knex, { offset, limit }))
      expect(users).toMatchSnapshot()
      offset += limit
      if (users.length < limit) {
        break
      }
    }
    done()
  })

  it('createUser without crashing', async (done) => {
    const { userId, userName } = await createUser(knex, user1)
    expect(userId).not.toBeNull()
    expect(userName).toBe(user1.userName.toLowerCase())
    done()
  })
  it('getUser by name successful', async (done) => {
    const { userName, email } = await getUserByNameOrEmail(knex, { userName: user1.userName })
    expect(userName).toBe(user1.userName.toLowerCase())
    expect(email).toBe(user1.email.toLowerCase())
    done()
  })
  it('getUser by email successful', async (done) => {
    const { userName, email } = await getUserByNameOrEmail(knex, { email: user1.email })
    expect(userName).toBe(user1.userName.toLowerCase())
    expect(email).toBe(user1.email.toLowerCase())
    done()
  })
  it('getUser by id successful', async (done) => {
    const { userId } = await getUserByNameOrEmail(knex, { userName: user1.userName })
    const user = await getUserById(knex, userId)
    expect(user.userName).toBe(user1.userName.toLowerCase())
    expect(user.email).toBe(user1.email.toLowerCase())
    done()
  })
  it('duplicate createUser to throw', async (done) => {
    await expect(createUser(knex, user1)).rejects.toThrow()
    done()
  })
  it('fails to create user with uuid formatted userName', async done => {
    await expect(createUser(knex, { userName: '129c4a2f-3c86-4c42-a175-908ee01176c7', email: 'test@test.com' })).rejects.toThrow()
    done()
  })
  it('fails to create user invalid email', async done => {
    await expect(createUser(knex, { userName: 'test2', email: 'invalidemail' })).rejects.toThrow()
    done()
  })
  it('updates user succeeds', async done => {
    const { userId } = await getUserByNameOrEmail(knex, { userName: user1.userName })
    const { count } = await updateUser(knex, userId, { email: 'newuser1@example.localhost' })
    expect(count).toBe(1)
    done()
  })
  it('remove user succeeds', async done => {
    const { userId } = await getUserByNameOrEmail(knex, { userName: user1.userName })
    const { count } = await removeUser(knex, userId)
    expect(count).toBe(1)
    done()
  })
})
