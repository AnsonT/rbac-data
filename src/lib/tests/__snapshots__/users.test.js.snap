// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Tests listUsers 1`] = `
Array [
  Object {
    "email": "Hattie_Hoeger@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "48234870-5389-445f-8b41-c61a52bf27dc",
    "userName": "Patrick36",
  },
  Object {
    "email": "Carson41@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "c5374da6-7722-46d3-a8ed-92ffe6d8b8b0",
    "userName": "Rodolfo_Renner",
  },
  Object {
    "email": "Waylon.Tillman@gmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "5abc0a5d-0ccf-450b-87af-bd8949dfb4d0",
    "userName": "Elliot_Borer",
  },
]
`;

exports[`Users Tests listUsers 2`] = `
Array [
  Object {
    "email": "Jaylon_Kulas@hotmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "58b20bce-a83c-4a7e-a28e-c5dbd448f2b7",
    "userName": "Jadyn.King88",
  },
  Object {
    "email": "Percy68@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "0cb50e5b-9e76-437a-afa7-8034f882c834",
    "userName": "Hillary_Hodkiewicz",
  },
  Object {
    "email": "Olin94@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "80a3bb8e-5916-45d0-ba4b-6cb353065773",
    "userName": "Amalia_Hauck72",
  },
]
`;

exports[`Users Tests listUsers 3`] = `
Array [
  Object {
    "email": "Norbert_Stark@hotmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "0b442035-ad75-4e79-b2d2-f506314a107b",
    "userName": "Fabiola0",
  },
  Object {
    "email": "Susanna_OReilly26@hotmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "c95c9225-de49-4bc9-9ee8-30d19f97d21a",
    "userName": "Maverick_Mueller",
  },
  Object {
    "email": "Johnpaul_Kozey@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "0bc7b4a0-638a-404e-bf54-7fa7f2088ba5",
    "userName": "Carmella_Glover37",
  },
]
`;

exports[`Users Tests listUsers 4`] = `
Array [
  Object {
    "email": "Mittie_Stokes@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "d7bda906-0f84-4c6b-b3e4-43a60a94b4fc",
    "userName": "Carmelo_Stark",
  },
  Object {
    "email": "Alfonso11@hotmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "d9dd08b0-b2bb-4c7e-8a42-bbbf9e1d00b0",
    "userName": "Frederique_Rohan92",
  },
  Object {
    "email": "Ethyl70@gmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "d2ed66c2-3a20-4411-99b4-959cd6b03a14",
    "userName": "Jevon39",
  },
]
`;

exports[`Users Tests listUsers 5`] = `
Array [
  Object {
    "email": "Luciano.Beer83@hotmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "0ea7988c-66fb-4df6-b69b-b72ed1b7bada",
    "userName": "Nels_Stokes64",
  },
  Object {
    "email": "Mathew.Renner73@gmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "b3ca82ad-e070-4309-a983-0806baba34c0",
    "userName": "Rosalind.Hegmann55",
  },
  Object {
    "email": "Rey83@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "a6ea8b9d-5508-4332-8885-48a60b10c570",
    "userName": "Wade69",
  },
]
`;

exports[`Users Tests listUsers 6`] = `
Array [
  Object {
    "email": "Tabitha.Ward@gmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "af23554a-0440-4a94-b7c6-6054ad9a4ab3",
    "userName": "Guadalupe.West4",
  },
  Object {
    "email": "Dina.Bayer@yahoo.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "7400349e-e892-4886-9dd9-2b33815f8003",
    "userName": "Donnie23",
  },
  Object {
    "email": "Johann11@gmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "9d9cc337-e0af-4009-b6e5-b38b8f44ea86",
    "userName": "Maude.Kemmer68",
  },
]
`;

exports[`Users Tests listUsers 7`] = `
Array [
  Object {
    "email": "Rozella45@gmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "e52b86e4-5f5e-4496-b7f1-e08dc8c52990",
    "userName": "Ayana50",
  },
  Object {
    "email": "Susie.Parker39@hotmail.com",
    "emailVerifiedAt": null,
    "needNewPassword": false,
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "userId": "3458bd53-c02e-4c58-94e5-cd1b6843ae97",
    "userName": "Tianna_Volkman",
  },
]
`;
